src/compute/src/sink/persist_sink.rs

```diff
@@ -1181,3 +1191,87 @@ where
+            match ac.frontier().to_owned().into_option() {
+                Some(ts) => {
+                    match refresh_schedule.round_up_timestamp(ts) {
+                        Some(rounded_up_ts) => {
+                            capability.as_mut().unwrap().downgrade(&rounded_up_ts);
+                        }
+                        None => {
+                            // We are past the last refresh. Drop the capability to signal that we
+                            // are done.
+                            capability = None;
+                        }
+                    }
+                }
+                None => {
+                    capability = None;
+                }
+            }
+            if let Some(capability) = &capability {
+                input.for_each(|_cap, data| {
+                    data.swap(&mut buffer);
+                    let mut cached_ts: Option<Timestamp> = None;
+                    let mut cached_rounded_up_ts = None;
+                    buffer
+                        .drain_filter_swapping(|(_d, ts, _r)| {
+                            let rounded_up_ts = {
+                                // We cache the rounded up timestamp for the last seen timestamp,
+                                // because the rounding up has a non-negligible cost. Caching for
+                                // just the 1 last timestamp helps already, because in some common
+                                // cases, we'll be seeing a lot of identical timestamps, e.g.,
+                                // during a rehydration, or when we have much more than 1000 records
+                                // during a single second.
+                                let some_ts_clone = Some(ts.clone());
+                                if cached_ts != some_ts_clone {
+                                    cached_ts = some_ts_clone;
+                                    cached_rounded_up_ts = refresh_schedule.round_up_timestamp(*ts);
+                                }
+                                cached_rounded_up_ts
+                            };
+                            match rounded_up_ts {
+                                Some(rounded_up_ts) => {
+                                    *ts = rounded_up_ts;
+                                    false
+                                }
+                                None => {
+                                    // This record is after the last refresh, so drop it.
+                                    true
+                                }
+                            }
+                        })
+                        .for_each(drop); // consume the iterator to apply the filtering
+                    output_buf
+                        .activate()
+                        .session(&capability)
+                        .give_container(&mut buffer);
+                });
+            }
+        }
+    });
+
+    use differential_dataflow::AsCollection;
```

Comment:
Nit: I'd prefer importing this trait once rather than at every call site.

Reply:
done